### ISSUE TITLE
GetListing() Object reference not set to an instance of an object.

### ISSUE THREAD

**Issue by RicardoJarree** [robinrodricks/FluentFTP#589]
**FTP OS:** Unix

**FTP Server:** Pure-FTPd

**Computer OS:** Windows 10

**FluentFTP Version:** 32.4.3

Everytime we try to get a listing of any directory on this server we keep getting a NullReferenceException. We have tried playing about with the settings, using AutoCorrect, different folders. This usually works for our other FTP's it's just this one which is having issues.

**Logs :**

```
# Connect()
Status:   Connecting to ******
Response: 220---------- Welcome to Pure-FTPd [privsep] [TLS] ----------
Response: 220-You are user number 1 of 50 allowed.
Response: 220-Local time is now 16:53. Server port: 21.
Response: 220-This is a private system - No anonymous login
Response: 220-IPv6 connections are also welcome on this server.
Response: 220 You will be disconnected after 15 minutes of inactivity.
Status:   Detected FTP server: PureFTPd
Command:  AUTH TLS
Response: 234 AUTH TLS OK.
Status:   FTPS Authentication Successful
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0.1889871.
Command:  USER *******
Response: 331 User ******** OK. Password required
Command:  PASS ***
Response: 230 OK. Current restricted directory is /
Command:  PBSZ 0
Response: 200 PBSZ=0
Command:  PROT P
Response: 200 Data protection level set to "private"
Command:  FEAT
Response: 211-Extensions supported:
Response: EPRT
Response: IDLE
Response: MDTM
Response: SIZE
Response: MFMT
Response: REST STREAM
Response: MLST type*;size*;sizd*;modify*;UNIX.mode*;UNIX.uid*;UNIX.gid*;unique*;
Response: MLSD
Response: AUTH TLS
Response: PBSZ
Response: PROT
Response: UTF8
Response: TVFS
Response: ESTA
Response: PASV
Response: EPSV
Response: SPSV
Response: ESTP
Response: 211 End.
Status:   Text encoding: System.Text.UTF8Encoding+UTF8EncodingSealed
Command:  OPTS UTF8 ON
Response: 200 OK, UTF-8 enabled
Command:  SYST
Response: 215 UNIX Type: L8
Status:   Testing connectivity using Socket.Poll()...

# GetListing("/public_html/magento/var/export_bkp", ForceList)
Status:   Testing connectivity using Socket.Poll()...
Command:  TYPE I
Response: 200 TYPE is now 8-bit binary

# OpenPassiveDataStream(AutoPassive, "LIST /public_html/magento/var/export_bkp", 0)
Command:  EPSV
Response: 229 Extended Passive mode OK (|||48102|)
Status:   Connecting to *********:48102
Command:  LIST /public_html/magento/var/export_bkp
Response: 150 Accepted data connection
Status:   FTPS Authentication Successful
Status:   Time to activate encryption: 0h 0m 0s.  Total Seconds: 0.0430484.
+---------------------------------------+
-----------------------------------------
Status:   Disposing FtpSocketStream...

# CloseDataStream()
Status:   Disposing FtpSocketStream...
An exception has been caught System.NullReferenceException: Object reference not set to an instance of an object.
   at FluentFTP.FtpClient.GetListingInternal(String listcmd, FtpListOption options, Boolean retry)
   at FluentFTP.FtpClient.GetListing(String path, FtpListOption options)
   at StcFtpTest.Program.Main(String[] args) in D:\Development\StcFtpTest\Program.cs:line 27

# Dispose()
Status:   Disposing FtpClient object...
Command:  QUIT
Status:   Disposing FtpSocketStream...
Status:   Disposing FtpSocketStream...

```

Code:
```
static void Main(string[] args)
        {
            FtpTrace.LogToConsole = true;
            using (FtpClient client = new FtpClient(******, 21, ********, ********))
            {
                client.EncryptionMode = FtpEncryptionMode.Explicit;
                client.RecursiveList = false;
                client.ValidateAnyCertificate = true;
                try
                {
                    client.Connect();
                    IList<FileInfo> fileInfos = new List<FileInfo>();
                    if (client.IsConnected)
                    {
                        var files = client.GetListing("/public_html/magento/var/export_bkp", FtpListOption.ForceList | FtpListOption.Auto);
                        foreach (var file in files)
                        {
                            Console.WriteLine(file);
                        }
                        client.Disconnect();
                    }
                    else
                    {
                        Console.WriteLine("Failed to connect to FTP.");
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine("An exception has been caught " + e.ToString());
                }
            }

            Console.ReadLine();
        }
```

**Comment by RicardoJarree** [comment]
Below are the logs from a connection to the same place on FileZilla.

```
17:13:05	Trace:	CFtpLogonOpData::Send() in state 0
17:13:05	Status:	Connecting to ********:21...
17:13:05	Status:	Connection established, waiting for welcome message...
17:13:05	Trace:	CFtpControlSocket::OnReceive()
17:13:05	Response:	220---------- Welcome to Pure-FTPd [privsep] [TLS] ----------
17:13:05	Response:	220-You are user number 1 of 50 allowed.
17:13:05	Response:	220-Local time is now 17:10. Server port: 21.
17:13:05	Response:	220-This is a private system - No anonymous login
17:13:05	Response:	220-IPv6 connections are also welcome on this server.
17:13:05	Response:	220 You will be disconnected after 15 minutes of inactivity.
17:13:05	Trace:	CFtpLogonOpData::ParseResponse() in state 1
17:13:05	Trace:	CControlSocket::SendNextCommand()
17:13:05	Trace:	CFtpLogonOpData::Send() in state 2
17:13:05	Command:	AUTH TLS
17:13:05	Trace:	CFtpControlSocket::OnReceive()
17:13:05	Response:	234 AUTH TLS OK.
17:13:05	Trace:	CFtpLogonOpData::ParseResponse() in state 2
17:13:05	Status:	Initializing TLS...
17:13:05	Trace:	tls_layer_impl::client_handshake()
17:13:05	Trace:	tls_layer_impl::continue_handshake()
17:13:05	Trace:	TLS handshake: About to send CLIENT HELLO
17:13:05	Trace:	TLS handshake: Sent CLIENT HELLO
17:13:05	Trace:	tls_layer_impl::on_send()
17:13:05	Trace:	tls_layer_impl::continue_handshake()
17:13:05	Trace:	tls_layer_impl::on_read()
17:13:05	Trace:	tls_layer_impl::continue_handshake()
17:13:05	Trace:	tls_layer_impl::on_read()
17:13:05	Trace:	tls_layer_impl::continue_handshake()
17:13:05	Trace:	TLS handshake: Received SERVER HELLO
17:13:05	Trace:	TLS handshake: Processed SERVER HELLO
17:13:05	Trace:	tls_layer_impl::on_read()
17:13:05	Trace:	tls_layer_impl::continue_handshake()
17:13:05	Trace:	TLS handshake: Received CERTIFICATE
17:13:05	Trace:	TLS handshake: Processed CERTIFICATE
17:13:05	Trace:	TLS handshake: Received SERVER KEY EXCHANGE
17:13:05	Trace:	TLS handshake: Processed SERVER KEY EXCHANGE
17:13:05	Trace:	TLS handshake: Received SERVER HELLO DONE
17:13:05	Trace:	TLS handshake: Processed SERVER HELLO DONE
17:13:06	Trace:	TLS handshake: About to send CLIENT KEY EXCHANGE
17:13:06	Trace:	TLS handshake: Sent CLIENT KEY EXCHANGE
17:13:06	Trace:	TLS handshake: About to send FINISHED
17:13:06	Trace:	TLS handshake: Sent FINISHED
17:13:06	Trace:	tls_layer_impl::on_read()
17:13:06	Trace:	tls_layer_impl::continue_handshake()
17:13:06	Trace:	TLS handshake: Received NEW SESSION TICKET
17:13:06	Trace:	TLS handshake: Processed NEW SESSION TICKET
17:13:06	Trace:	TLS handshake: Received FINISHED
17:13:06	Trace:	TLS handshake: Processed FINISHED
17:13:06	Trace:	TLS Handshake successful
17:13:06	Trace:	Protocol: TLS1.2, Key exchange: ECDHE-SECP256R1-RSA-SHA512, Cipher: AES-256-GCM, MAC: AEAD
17:13:06	Trace:	tls_layer_impl::verify_certificate()
17:13:06	Status:	Verifying certificate...
17:13:06	Trace:	CFtpControlSocket::SetAsyncRequestReply
17:13:06	Status:	TLS connection established.
17:13:06	Trace:	CControlSocket::SendNextCommand()
17:13:06	Trace:	CFtpLogonOpData::Send() in state 6
17:13:06	Command:	USER ********
17:13:06	Trace:	CFtpControlSocket::OnReceive()
17:13:06	Trace:	tls_layer_impl::on_read()
17:13:06	Trace:	CFtpControlSocket::OnReceive()
17:13:06	Response:	331 User ******** OK. Password required
17:13:06	Trace:	CFtpLogonOpData::ParseResponse() in state 6
17:13:06	Trace:	CControlSocket::SendNextCommand()
17:13:06	Trace:	CFtpLogonOpData::Send() in state 6
17:13:06	Command:	PASS **********
17:13:06	Trace:	tls_layer_impl::on_read()
17:13:06	Trace:	CFtpControlSocket::OnReceive()
17:13:06	Response:	230 OK. Current restricted directory is /
17:13:06	Trace:	CFtpLogonOpData::ParseResponse() in state 6
17:13:06	Trace:	CControlSocket::SendNextCommand()
17:13:06	Trace:	CFtpLogonOpData::Send() in state 10
17:13:06	Command:	OPTS UTF8 ON
17:13:06	Trace:	tls_layer_impl::on_read()
17:13:06	Trace:	CFtpControlSocket::OnReceive()
17:13:06	Response:	200 OK, UTF-8 enabled
17:13:06	Trace:	CFtpLogonOpData::ParseResponse() in state 10
17:13:06	Trace:	CControlSocket::SendNextCommand()
17:13:06	Trace:	CFtpLogonOpData::Send() in state 11
17:13:06	Command:	PBSZ 0
17:13:06	Trace:	tls_layer_impl::on_read()
17:13:06	Trace:	CFtpControlSocket::OnReceive()
17:13:06	Response:	200 PBSZ=0
17:13:06	Trace:	CFtpLogonOpData::ParseResponse() in state 11
17:13:06	Trace:	CControlSocket::SendNextCommand()
17:13:06	Trace:	CFtpLogonOpData::Send() in state 12
17:13:06	Command:	PROT P
17:13:06	Trace:	tls_layer_impl::on_read()
17:13:06	Trace:	CFtpControlSocket::OnReceive()
17:13:06	Response:	200 Data protection level set to "private"
17:13:06	Trace:	CFtpLogonOpData::ParseResponse() in state 12
17:13:06	Status:	Logged in
17:13:06	Trace:	Measured latency of 19 ms
17:13:06	Trace:	CFtpControlSocket::ResetOperation(0)
17:13:06	Trace:	CControlSocket::ResetOperation(0)
17:13:06	Trace:	CFtpLogonOpData::Reset(0) in state 15
17:13:06	Trace:	CFileZillaEnginePrivate::ResetOperation(0)
17:13:06	Trace:	CFileZillaEnginePrivate::ResetOperation(0)
17:15:28	Trace:	CControlSocket::SendNextCommand()
17:15:28	Trace:	CFtpListOpData::Send() in state 0
17:15:28	Status:	Retrieving directory listing of "/public_html/magento/var/export_bkp"...
17:15:28	Trace:	CFtpChangeDirOpData::Send() in state 0
17:15:28	Trace:	CFtpChangeDirOpData::Send() in state 2
17:15:28	Command:	CWD /public_html/magento/var/export_bkp
17:15:28	Trace:	tls_layer_impl::on_read()
17:15:28	Trace:	CFtpControlSocket::OnReceive()
17:15:28	Response:	250 OK. Current directory is /public_html/magento/var/export_bkp
17:15:28	Trace:	CFtpChangeDirOpData::ParseResponse() in state 2
17:15:28	Trace:	CFtpControlSocket::ResetOperation(0)
17:15:28	Trace:	CControlSocket::ResetOperation(0)
17:15:28	Trace:	CFtpChangeDirOpData::Reset(0) in state 2
17:15:28	Trace:	CFtpListOpData::SubcommandResult(0) in state 1
17:15:28	Trace:	CControlSocket::SendNextCommand()
17:15:28	Trace:	CFtpListOpData::Send() in state 2
17:15:28	Trace:	CFtpRawTransferOpData::Send() in state 0
17:15:28	Trace:	CFtpRawTransferOpData::Send() in state 2
17:15:28	Command:	PASV
17:15:28	Trace:	tls_layer_impl::on_read()
17:15:28	Trace:	CFtpControlSocket::OnReceive()
17:15:28	Response:	227 Entering Passive Mode (5,134,10,249,179,216)
17:15:28	Trace:	CFtpRawTransferOpData::ParseResponse() in state 2
17:15:28	Trace:	CControlSocket::SendNextCommand()
17:15:28	Trace:	CFtpRawTransferOpData::Send() in state 4
17:15:28	Trace:	Binding data connection source IP to control connection source IP 10.1.2.28
17:15:28	Trace:	tls_layer_impl::client_handshake()
17:15:28	Trace:	Trying to resume existing TLS session.
17:15:28	Command:	MLSD
17:15:28	Trace:	tls_layer_impl::on_send()
17:15:28	Trace:	tls_layer_impl::continue_handshake()
17:15:28	Trace:	TLS handshake: About to send CLIENT HELLO
17:15:28	Trace:	TLS handshake: Sent CLIENT HELLO
17:15:28	Trace:	tls_layer_impl::on_read()
17:15:28	Trace:	tls_layer_impl::on_read()
17:15:28	Trace:	tls_layer_impl::continue_handshake()
17:15:28	Trace:	TLS handshake: Received SERVER HELLO
17:15:28	Trace:	TLS handshake: Processed SERVER HELLO
17:15:28	Trace:	TLS handshake: Received FINISHED
17:15:28	Trace:	TLS handshake: Processed FINISHED
17:15:28	Trace:	TLS handshake: About to send FINISHED
17:15:28	Trace:	TLS handshake: Sent FINISHED
17:15:28	Trace:	TLS Handshake successful
17:15:28	Trace:	TLS Session resumed
17:15:28	Trace:	Protocol: TLS1.2, Key exchange: ECDHE-SECP256R1, Cipher: AES-256-GCM, MAC: AEAD
17:15:28	Trace:	tls_layer_impl::verify_certificate()
17:15:28	Trace:	CFtpControlSocket::OnReceive()
17:15:28	Response:	150 Accepted data connection
17:15:28	Trace:	CFtpRawTransferOpData::ParseResponse() in state 4
17:15:28	Trace:	CControlSocket::SendNextCommand()
17:15:28	Trace:	CFtpRawTransferOpData::Send() in state 5
17:15:28	Trace:	CTransferSocket::OnConnect
17:15:28	Trace:	CTransferSocket::OnReceive(), m_transferMode=0
17:15:28	Trace:	tls_layer_impl::on_read()
17:15:28	Trace:	CTransferSocket::OnReceive(), m_transferMode=0
17:15:28	Trace:	tls_layer_impl::on_read()
17:15:28	Trace:	CTransferSocket::OnReceive(), m_transferMode=0
17:15:28	Trace:	CTransferSocket::TransferEnd(1)
17:15:28	Trace:	tls_layer_impl::shutdown()
17:15:28	Trace:	tls_layer_impl::continue_shutdown()
17:15:28	Trace:	CFtpControlSocket::TransferEnd()
17:15:28	Trace:	tls_layer_impl::on_read()
17:15:28	Trace:	CFtpControlSocket::OnReceive()
17:15:28	Response:	226-Options: -a -l 
17:15:28	Response:	226 3 matches total
17:15:28	Trace:	CFtpRawTransferOpData::ParseResponse() in state 7
17:15:28	Trace:	CFtpControlSocket::ResetOperation(0)
17:15:28	Trace:	CControlSocket::ResetOperation(0)
17:15:28	Trace:	CFtpRawTransferOpData::Reset(0) in state 7
17:15:28	Trace:	CFtpListOpData::SubcommandResult(0) in state 3
17:15:28	Listing:	type=cdir;sizd=4096;modify=20200618152126;UNIX.mode=0777;UNIX.uid=1006;UNIX.gid=1009;unique=fd01gafa414; .
17:15:28	Listing:	type=pdir;sizd=4096;modify=20190705131437;UNIX.mode=0755;UNIX.uid=1006;UNIX.gid=1009;unique=fd01gad03f8; ..
17:15:29	Listing:	type=file;size=0;modify=20200618152126;UNIX.mode=0644;UNIX.uid=1006;UNIX.gid=1009;unique=fd01gae6c08; test.txt
17:15:29	Trace:	CFtpControlSocket::ResetOperation(0)
17:15:29	Trace:	CControlSocket::ResetOperation(0)
17:15:29	Trace:	CFtpListOpData::Reset(0) in state 3
17:15:29	Status:	Directory listing of "/public_html/magento/var/export_bkp" successful
17:15:29	Trace:	CFileZillaEnginePrivate::ResetOperation(0)
```

**Comment by robinrodricks** [comment]
Can you try GetListing with the path of the filename.

**Comment by robinrodricks** [comment]
Also, any  reason you are using FtpListOption.ForceList? this will force the older LIST command.

**Comment by robinrodricks** [comment]
Since its a simple null exception, can you build from source and see where the exception occurs? Its impossible for me to debug this on your setup.

**Comment by RicardoJarree** [comment]
Okay so here is where it first stops, there is no completion code and I know for a fact the folder has a txt file in WinSCP and FileZilla:

![Annotation 2020-06-30 233913](https://user-images.githubusercontent.com/5776374/86183965-12b41100-bb2b-11ea-9981-3e47c9e61b0a.png)

This is the stack trace:

![image](https://user-images.githubusercontent.com/5776374/86184069-45f6a000-bb2b-11ea-8588-2dc0c859658f.png)

In FtpClient_Stream.cs it's failing at the below point as it seems to think there isn't a return..?

![image](https://user-images.githubusercontent.com/5776374/86184253-b3a2cc00-bb2b-11ea-924c-6e587f00cf01.png)

I'll probably continue looking further into this myself tomorrow, not sure if anything jumps out at you @robinrodricks based on my previous logs?

**Comment by RicardoJarree** [comment]
Aha! Turns out I was in the perfect place to work out the solution, I needed to turn on the _FtpListOption.UseStat_ option! Cheers @robinrodricks for pointing me in the right direction!

**Comment by robinrodricks** [comment]
I applied a fix anyways in 33.0.0.
